java/src/main/native/src/CuFileJni.cpp

```diff
@@ -0,0 +1,237 @@
+    if (status.err != CU_FILE_SUCCESS) {
+      close(file_descriptor_);
+      throw std::logic_error("Failed to register cuFile handle: " + cuFileGetErrorString(status));
+    }
+  }
+
+  /**
+   * @brief Factory method to create a file wrapper for reading.
+   *
+   * @param path Absolute path of the file to read from. This file must exist.
+   * @return std::unique_ptr<cufile_file> for reading.
+   */
+  static auto make_reader(char const* path)
+  {
+    auto const file_descriptor = open(path, O_RDONLY | O_DIRECT);
+    if (file_descriptor < 0) {
+      throw std::logic_error("Failed to open file to read: " + cuFileGetErrorString(errno));
+    }
+    return std::make_unique<cufile_file>(file_descriptor);
+  }
+
+  /**
+   * @brief Factory method to create a file wrapper for writing.
+   *
+   * @param path Absolute path of the file to write to. This creates the file if it does not already
+   * exist..
+   * @return std::unique_ptr<cufile_file> for writing.
+   */
+  static auto make_writer(char const* path)
+  {
+    auto const file_descriptor = open(path, O_CREAT | O_WRONLY | O_DIRECT, S_IRUSR | S_IWUSR);
+    if (file_descriptor < 0) {
+      throw std::logic_error("Failed to open file to write: " + cuFileGetErrorString(errno));
+    }
+    return std::make_unique<cufile_file>(file_descriptor);
+  }
+
+  // Disable copy (and move) semantics.
+  cufile_file(cufile_file const&)            = delete;
+  cufile_file& operator=(cufile_file const&) = delete;
+
+  /** @brief Destroy the file wrapper by de-registering the cuFile handle and closing the file. */
+  ~cufile_file()
+  {
+    cuFileHandleDeregister(cufile_handle_);
+    close(file_descriptor_);
+  }
+
+  /**
+   * @brief Read the file into a device buffer.
+   *
+   * @param buffer Device buffer to read the file content into.
+   * @param file_offset Starting offset from which to read the file.
+   */
+  void read(void* buffer,
+            std::size_t size,
+            std::size_t file_offset,
+            std::size_t device_offset) const
+  {
+    auto const status = cuFileRead(cufile_handle_, buffer, size, file_offset, device_offset);
+
+    if (status < 0) {
+      if (IS_CUFILE_ERR(status)) {
+        throw std::logic_error("Failed to read file into buffer: " + cuFileGetErrorString(status));
+      } else {
+        throw std::logic_error("Failed to read file into buffer: " + cuFileGetErrorString(errno));
+      }
+    }
+  }
+
+  void write(void* buffer, std::size_t size, std::size_t file_offset, std::size_t buffer_offset)
+  {
+    auto const status = cuFileWrite(cufile_handle_, buffer, size, file_offset, buffer_offset);
+    if (status < 0) {
+      if (IS_CUFILE_ERR(status)) {
+        throw std::logic_error("Failed to write file from buffer: " + cuFileGetErrorString(status));
+      } else {
+        throw std::logic_error("Failed to write file from buffer: " + cuFileGetErrorString(errno));
+      }
+    }
+  }
```

Comment:
Consider adding a `sync_default_stream` argument. 
>   Contrary to most of the non-async CUDA API, cuFile does not have the semantic of being ordered
>   with respect to other non-cuFile work in the default stream. By enabling `sync_default_stream`,
>   KvikIO will synchronize the default stream and order the operation with respect to other work
>   in the null stream. When in KvikIO's compatibility mode or when accessing host memory, the
>   operation is always default stream ordered like the rest of the non-async CUDA API. In this
>   case, the value of `sync_default_stream` is ignored.

See https://github.com/rapidsai/kvikio/blob/branch-24.10/cpp/include/kvikio/file_handle.hpp#L317-L323

Reply:
Hey Mads, just saw this. I think this is a good suggestion and one we should definitely incorporate. My bandwidth is very limited at the moment so I am unable to add it now, would you be okay with adding this in a future update? I think this will require some updates to the maven file to support the proper compilation with the cuda libraries and it takes me a good amount of time to context switch+remember how maven works.